File: server/shared/typescript/encryption/src/ecdh-p256.ts
// -----------------------------------------------------------------------
// Copyright (c) DCSV. All rights reserved.
// -----------------------------------------------------------------------

import { subtle } from "./subtle.js";
import type { WebCryptoKey } from "./subtle.js";

/**
 * Import + agreement helpers for the P-256 ECDH keys the sealed encryption
 * mode uses — the WebCrypto twin of .NET `EcdhP256`.
 *
 * All imports are strict: WebCrypto `importKey` with `namedCurve: "P-256"`
 * validates the DER structure AND that the point lies on the P-256 curve
 * (a wrong-curve or invalid-point key is rejected with a `DataError`). So the
 * WebCrypto import is BOTH the structural import and the functional agreement
 * probe the .NET twin runs as two steps — callers wrap the rejection into
 * their surface-appropriate error.
 */

/** Raw P-256 shared-secret length in bytes (the 32-byte X-coordinate). */
const _SHARED_SECRET_BYTES = 32;

/** ECDH raw-secret length in bits, for `deriveBits`. */
const _SHARED_SECRET_BITS = _SHARED_SECRET_BYTES * 8;

const _P256: EcKeyImportParams = { name: "ECDH", namedCurve: "P-256" };

/**
 * Imports a SubjectPublicKeyInfo public key as an ECDH P-256 public key.
 *
 * @param spki SubjectPublicKeyInfo DER bytes.
 * @returns The imported public key. A public key carries no key usages
 *   (`deriveBits` names it in the algorithm, not as the operating key).
 * @throws On DER-invalid, wrong-curve, or off-curve material (WebCrypto error).
 */
export function importPublicP256(spki: Uint8Array): Promise<WebCryptoKey> {
  return subtle.importKey("spki", spki, _P256, true, []);
}

/**
 * Imports a PKCS#8 private key as an ECDH P-256 private key.
 *
 * @param pkcs8 PKCS#8 PrivateKeyInfo DER bytes.
 * @returns The imported private key (usable for `deriveBits`).
 * @throws On DER-invalid, public-only, wrong-curve, or off-curve material.
 */
export function importPrivateP256(pkcs8: Uint8Array): Promise<WebCryptoKey> {
  return subtle.importKey("pkcs8", pkcs8, _P256, false, ["deriveBits"]);
}

/** A freshly generated ephemeral P-256 keypair for one seal. */
export interface EphemeralP256Keypair {
  /** The private half — used once to derive the shared secret, then dropped. */
  readonly privateKey: WebCryptoKey;
  /** The public half as SubjectPublicKeyInfo DER — travels in the sealed frame. */
  readonly publicSpki: Uint8Array;
}

/**
 * Generates a fresh ephemeral P-256 keypair and exports its public half as
 * SPKI DER (the bytes that ride in the sealed frame and bind into the HKDF
 * info). The twin of the per-seal `ECDiffieHellman.Create(nistP256)` +
 * `ExportSubjectPublicKeyInfo()` in the .NET sealer.
 *
 * @returns The ephemeral private key + its exported public SPKI.
 */
export async function generateEphemeralP256(): Promise<EphemeralP256Keypair> {
  const pair = await subtle.generateKey(_P256, true, ["deriveBits"]);
  const publicSpki = await subtle.exportKey("spki", pair.publicKey);

  return {
    privateKey: pair.privateKey,
    publicSpki: new Uint8Array(publicSpki),
  };
}

/**
 * Derives the raw P-256 ECDH shared secret (the 32-byte X-coordinate of the
 * agreed point) between a private key and a peer public key. Byte-identical to
 * .NET `ECDiffieHellman.DeriveRawSecretAgreement(peerPublic)` — KAT-pinned.
 *
 * @param privateKey This side's private key.
 * @param peerPublicKey The peer's public key.
 * @returns The 32-byte raw shared secret.
 */
export async function deriveRawSecret(
  privateKey: WebCryptoKey,
  peerPublicKey: WebCryptoKey,
): Promise<Uint8Array> {
  const secret = await subtle.deriveBits(
    { name: "ECDH", public: peerPublicKey },
    privateKey,
    _SHARED_SECRET_BITS,
  );

  return new Uint8Array(secret);
}

File: public/packages/dotnet/encryption/core/EcdhP256.cs
// -----------------------------------------------------------------------
// <copyright file="EcdhP256.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Encryption;

using System.Security.Cryptography;
using JetBrains.Annotations;

/// <summary>
/// Import + validation helpers for the P-256 ECDH keys the sealed encryption
/// mode uses. All imports are strict: trailing bytes after the DER structure,
/// a non-256-bit curve, or key material that cannot complete a P-256
/// agreement are rejected with <see cref="CryptographicException"/> — callers
/// wrap into their surface-appropriate exception (keyring constructors throw
/// <see cref="ArgumentException"/>; the opener throws
/// <see cref="FrameMalformedException"/> for frame-borne keys).
/// </summary>
internal static class EcdhP256
{
    /// <summary>P-256 key size in bits.</summary>
    internal const int KEY_SIZE_BITS = 256;

    /// <summary>
    /// Imports a SubjectPublicKeyInfo public key, requiring a 256-bit curve.
    /// </summary>
    /// <param name="spki">SubjectPublicKeyInfo DER bytes.</param>
    /// <returns>The imported key. Caller disposes.</returns>
    [MustDisposeResource]
    internal static ECDiffieHellman ImportPublicP256(ReadOnlySpan<byte> spki)
    {
        var ecdh = ECDiffieHellman.Create();

        try
        {
            try
            {
                ecdh.ImportSubjectPublicKeyInfo(spki, out var bytesRead);

                if (bytesRead != spki.Length)
                {
                    throw new CryptographicException(
                        "Trailing bytes after the SubjectPublicKeyInfo structure.");
                }
            }
            catch (PlatformNotSupportedException ex)
            {
                // Windows CNG raises PlatformNotSupportedException for a
                // DER-valid SPKI whose point is invalid for the declared
                // curve. Normalize: it is bad KEY MATERIAL, not a missing
                // platform capability.
                throw new CryptographicException(
                    "Key material is not a valid P-256 public key.", ex);
            }

            if (ecdh.KeySize != KEY_SIZE_BITS)
            {
                throw new CryptographicException(
                    $"Key size {ecdh.KeySize} is not the required {KEY_SIZE_BITS} (P-256).");
            }

            return ecdh;
        }
        catch
        {
            ecdh.Dispose();
            throw;
        }
    }

    /// <summary>
    /// Imports a PKCS#8 private key, requiring a 256-bit curve.
    /// </summary>
    /// <param name="pkcs8">PKCS#8 PrivateKeyInfo DER bytes.</param>
    /// <returns>The imported key. Caller disposes.</returns>
    [MustDisposeResource]
    internal static ECDiffieHellman ImportPrivatePkcs8P256(ReadOnlySpan<byte> pkcs8)
    {
        var ecdh = ECDiffieHellman.Create();

        try
        {
            try
            {
                ecdh.ImportPkcs8PrivateKey(pkcs8, out var bytesRead);

                if (bytesRead != pkcs8.Length)
                {
                    throw new CryptographicException(
                        "Trailing bytes after the PKCS#8 PrivateKeyInfo structure.");
                }
            }
            catch (PlatformNotSupportedException ex)
            {
                // See ImportPublicP256 — normalize the CNG invalid-material
                // shape to the cryptographic failure it actually is.
                throw new CryptographicException(
                    "Key material is not a valid P-256 private key.", ex);
            }

            if (ecdh.KeySize != KEY_SIZE_BITS)
            {
                throw new CryptographicException(
                    $"Key size {ecdh.KeySize} is not the required {KEY_SIZE_BITS} (P-256).");
            }

            return ecdh;
        }
        catch
        {
            ecdh.Dispose();
            throw;
        }
    }

    /// <summary>
    /// Proves the key can complete a real P-256 ECDH agreement — a functional
    /// check that catches same-size non-P-256 curves and invalid points that
    /// survive structural import. Construction-time only (each probe costs a
    /// throwaway keygen + one agreement); the per-message paths rely on the
    /// real derivation as their probe.
    /// </summary>
    /// <param name="key">The imported key to probe.</param>
    internal static void ProbeP256Agreement(ECDiffieHellman key)
    {
        using var probe = ECDiffieHellman.Create(ECCurve.NamedCurves.nistP256);
        byte[]? sharedSecret = null;

        try
        {
            using var publicPart = key.PublicKey;
            sharedSecret = probe.DeriveRawSecretAgreement(publicPart);
        }
        catch (ArgumentException ex)
        {
            throw new CryptographicException(
                "Key cannot complete a P-256 ECDH agreement (wrong curve).", ex);
        }
        finally
        {
            if (sharedSecret is not null)
                CryptographicOperations.ZeroMemory(sharedSecret);
        }
    }
}
